File: backend/app/api/v1/endpoints/agent_text_units.py
"""Agent text-unit API — the fleet's ``.claude/`` corpus as one kind-discriminated
store of text-file bundles.

A unit is ``(kind, name)`` plus a ``files`` map of relative path → text.
``kind=command`` is the degenerate single-file case (``<name>.md``);
``kind=skill`` carries ``SKILL.md`` plus siblings. ``kind`` is a widenable
discriminator, not a two-value enum.

Two layers resolve here, ``account override → fleet default``:

* Every route is scoped to ONE organization by default. The org is either named
  explicitly (``organization_id``) and checked with the existing
  ``check_organization_membership``, or defaults to the caller's personal
  organization.
* ``fleet_default=true`` addresses the ``organization_id IS NULL`` layer
  instead. **Reads of it are open** to any authenticated caller — a fleet
  default is fleet-wide by definition — while **writes require a superuser**,
  because one write there changes every account that has not overridden the
  unit.

There is still no row for the runner's *embedded* default, so ``DELETE`` only
ever removes a stored layer and lets the next one down apply.

Two list projections, one query
-------------------------------

``GET ""`` returns whole units; ``GET "/index"`` returns the same rows with
every ``files`` map replaced by its paths, byte count and digest. The split is
not an optimization we reached for — the runner resolves the corpus on the
**spawn critical path** inside a 4 s budget and fails SOFT, degrading to cache
and then to embedded defaults, so a corpus too large to fetch in time does not
error, it silently disappears. Measured over the fleet corpus the full listing
is 1,988,661 bytes and the index is 47,093 (2.4%); the intended exchange is read
the index, diff ``checksum`` against the cache, then re-read ``GET ""`` with
``names=`` for just the units that moved. Both routes take the identical filter
set (``UnitListFilters``) and run the identical query, so the index can never
describe a listing the listing would not return.
"""

from typing import cast
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import current_active_user, get_async_db
from app.models.user import User
from app.services.agent_text_unit_service import (
    MAX_NAMES_PER_QUERY,
    AgentTextUnitCreate,
    AgentTextUnitIndexResponse,
    AgentTextUnitListResponse,
    AgentTextUnitResponse,
    AgentTextUnitService,
    AgentTextUnitUpdate,
    AgentTextUnitValidationError,
    AgentTextUnitVersionListResponse,
    RevertRequest,
)
from app.services.permissions.organization_access import (
    check_organization_membership,
    get_personal_organization,
)

router = APIRouter()


def get_service() -> AgentTextUnitService:
    return AgentTextUnitService()


async def resolve_layer(
    db: AsyncSession,
    user: User,
    organization_id: UUID | None,
    fleet_default: bool,
    *,
    write: bool,
) -> UUID | None:
    """Resolve + authorize the LAYER this request addresses.

    Returns ``None`` for the fleet-default layer and a UUID for an account's
    own. A fleet-default WRITE needs a superuser: the row it touches is
    inherited by every account that has not overridden the unit, so account
    membership is not the right authority for it.
    """
    if fleet_default:
        if write and not user.is_superuser:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Writing a fleet default requires a superuser",
            )
        return None

    if organization_id is not None:
        membership = await check_organization_membership(
            db, user.id, organization_id, required_role="member"
        )
        if not membership:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Not a member of this organization",
            )
        return organization_id

    personal = await get_personal_organization(db, user.id)
    if not personal:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=(
                "No personal organization for this user; pass organization_id "
                "explicitly."
            ),
        )
    # `Organization` is a legacy `Column`-style model, so `.id` types as a
    # Column rather than a UUID.
    return cast(UUID, personal.id)


def _validation_error(exc: AgentTextUnitValidationError) -> HTTPException:
    return HTTPException(
        status_code=status.HTTP_422_UNPROCESSABLE_CONTENT, detail=str(exc)
    )


class UnitListFilters:
    """The filter set BOTH list projections accept, declared exactly once.

    A second hand-written copy of these eight parameters is precisely how the
    full listing and the index would drift apart — one growing a filter the
    other lacks, so a client's index no longer describes the listing it is
    caching against. One dependency is how they cannot; FastAPI flattens it into
    each operation's parameters, so the published OpenAPI is unchanged by the
    sharing.
    """

    def __init__(
        self,
        kind: str | None = Query(
            None, description="Filter to one kind, e.g. 'command'"
        ),
        organization_id: UUID | None = Query(None),
        fleet_default: bool = Query(False, description="List the fleet-default layer"),
        include_fleet_defaults: bool = Query(
            True,
            description=(
                "Include fleet defaults the account has not overridden — the "
                "RESOLVED view the runner needs"
            ),
        ),
        invocable_only: bool = Query(
            False,
            description=(
                "Drop the non-invocable units (the underscore-prefixed copy-source "
                "specs). A client that PROVISIONS units to disk must pass true: a "
                "`_gate-registration.md` written into `.claude/commands/` becomes an "
                "invocable slash command. A client that EDITS the corpus wants the "
                "default, false, so the specs stay visible."
            ),
        ),
        names: list[str] | None = Query(
            None,
            description=(
                "Restrict the listing to these unit names. This is how a client "
                "that already holds the index fetches bodies for exactly the units "
                "whose `checksum` moved, in one request rather than one round trip "
                "per unit. Names are matched inside the `kind` filter, so pass "
                f"`kind` too when a name exists under more than one. At most "
                f"{MAX_NAMES_PER_QUERY}."
            ),
        ),
        offset: int = Query(0, ge=0),
        limit: int = Query(100, ge=1, le=500),
    ) -> None:
        self.kind = kind
        self.organization_id = organization_id
        self.fleet_default = fleet_default
        self.include_fleet_defaults = include_fleet_defaults
        self.invocable_only = invocable_only
        self.names = names
        self.offset = offset
        self.limit = limit


@router.get(
    "/index",
    response_model=AgentTextUnitIndexResponse,
    summary="List the agent text units WITHOUT their bodies",
)
async def list_agent_text_unit_index(
    filters: UnitListFilters = Depends(),
    db: AsyncSession = Depends(get_async_db),
    current_user: User = Depends(current_active_user),
    service: AgentTextUnitService = Depends(get_service),
) -> AgentTextUnitIndexResponse:
    """The resolved view with every `files` map replaced by its digest.

    Same rows, same filters, same pagination as `GET ""` — see
    `AgentTextUnitService.list_unit_index`, which shares the query rather than
    re-deriving it. Measured over the fleet corpus (87 units, 2026-08-25) this
    is 47,093 bytes against the full listing's 1,988,661 — 2.4% — which is what
    puts a cold resolve back
    inside the runner's 4 s spawn-path fetch budget.

    The intended exchange is two calls: read this index, diff `checksum` against
    the cache, then `GET ""?names=…` for the units that moved.

    Declared BEFORE `GET /{name}`, which would otherwise match `/index`. The
    cost is that a unit literally named `index` is not addressable through that
    route — it is still listed, editable and provisioned like any other, and the
    codebase already accepts this trade for `/search` and `/stats` elsewhere.
    """
    layer = await resolve_layer(
        db, current_user, filters.organization_id, filters.fleet_default, write=False
    )
    try:
        return await service.list_unit_index(
            db,
            layer,
            kind=filters.kind,
            include_fleet_defaults=filters.include_fleet_defaults,
            invocable_only=filters.invocable_only,
            names=filters.names,
            offset=filters.offset,
            limit=filters.limit,
        )
    except AgentTextUnitValidationError as exc:
        raise _validation_error(exc) from exc


@router.get(
    "",
    response_model=AgentTextUnitListResponse,
    summary="List the agent text units this account resolves to",
)
async def list_agent_text_units(
    filters: UnitListFilters = Depends(),
    db: AsyncSession = Depends(get_async_db),
    current_user: User = Depends(current_active_user),
    service: AgentTextUnitService = Depends(get_service),
) -> AgentTextUnitListResponse:
    """The resolved view: account overrides, plus unshadowed fleet defaults."""
    layer = await resolve_layer(
        db, current_user, filters.organization_id, filters.fleet_default, write=False
    )
    try:
        return await service.list_units(
            db,
            layer,
            kind=filters.kind,
            include_fleet_defaults=filters.include_fleet_defaults,
            invocable_only=filters.invocable_only,
            names=filters.names,
            offset=filters.offset,
            limit=filters.limit,
        )
    except AgentTextUnitValidationError as exc:
        raise _validation_error(exc) from exc


@router.post(
    "",
    response_model=AgentTextUnitResponse,
    summary="Create or replace an agent text unit",
)
async def upsert_agent_text_unit(
    data: AgentTextUnitCreate,
    organization_id: UUID | None = Query(None),
    fleet_default: bool = Query(False, description="Write the fleet-default layer"),
    db: AsyncSession = Depends(get_async_db),
    current_user: User = Depends(current_active_user),
    service: AgentTextUnitService = Depends(get_service),
) -> AgentTextUnitResponse:
    """Upsert by ``(layer, kind, name)``; appends a new version."""
    layer = await resolve_layer(
        db, current_user, organization_id, fleet_default, write=True
    )
    try:
        return await service.upsert_unit(db, layer, data, current_user.id)
    except AgentTextUnitValidationError as exc:
        raise _validation_error(exc) from exc


@router.get(
    "/{name}",
    response_model=AgentTextUnitResponse,
    summary="Get one agent text unit",
)
async def get_agent_text_unit(
    name: str,
    kind: str = Query(..., description="The unit kind, e.g. 'command' or 'skill'"),
    organization_id: UUID | None = Query(None),
    fleet_default: bool = Query(False),
    include_fleet_defaults: bool = Query(True),
    db: AsyncSession = Depends(get_async_db),
    current_user: User = Depends(current_active_user),
    service: AgentTextUnitService = Depends(get_service),
) -> AgentTextUnitResponse:
    """Resolves ``account override → fleet default``."""
    layer = await resolve_layer(
        db, current_user, organization_id, fleet_default, write=False
    )
    try:
        return await service.get_unit(
            db, layer, kind, name, include_fleet_defaults=include_fleet_defaults
        )
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)
        ) from exc


@router.patch(
    "/{name}",
    response_model=AgentTextUnitResponse,
    summary="Update an agent text unit",
)
async def update_agent_text_unit(
    name: str,
    data: AgentTextUnitUpdate,
    kind: str = Query(...),
    organization_id: UUID | None = Query(None),
    fleet_default: bool = Query(False),
    db: AsyncSession = Depends(get_async_db),
    current_user: User = Depends(current_active_user),
    service: AgentTextUnitService = Depends(get_service),
) -> AgentTextUnitResponse:
    """A ``files`` change appends a version; other fields do not."""
    layer = await resolve_layer(
        db, current_user, organization_id, fleet_default, write=True
    )
    try:
        return await service.update_unit(db, layer, kind, name, data, current_user.id)
    except AgentTextUnitValidationError as exc:
        raise _validation_error(exc) from exc
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)
        ) from exc


@router.get(
    "/{name}/versions",
    response_model=AgentTextUnitVersionListResponse,
    summary="List a unit's version history",
)
async def list_agent_text_unit_versions(
    name: str,
    kind: str = Query(...),
    organization_id: UUID | None = Query(None),
    fleet_default: bool = Query(False),
    offset: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500),
    db: AsyncSession = Depends(get_async_db),
    current_user: User = Depends(current_active_user),
    service: AgentTextUnitService = Depends(get_service),
) -> AgentTextUnitVersionListResponse:
    """History belongs to the addressed layer only — never a fleet fallback."""
    layer = await resolve_layer(
        db, current_user, organization_id, fleet_default, write=False
    )
    try:
        return await service.list_versions(db, layer, kind, name, offset, limit)
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)
        ) from exc


@router.post(
    "/{name}/revert",
    response_model=AgentTextUnitResponse,
    summary="Revert a unit to an earlier version",
)
async def revert_agent_text_unit(
    name: str,
    data: RevertRequest,
    kind: str = Query(...),
    organization_id: UUID | None = Query(None),
    fleet_default: bool = Query(False),
    db: AsyncSession = Depends(get_async_db),
    current_user: User = Depends(current_active_user),
    service: AgentTextUnitService = Depends(get_service),
) -> AgentTextUnitResponse:
    """Appends a NEW version whose files equal the target's — never a rewrite."""
    layer = await resolve_layer(
        db, current_user, organization_id, fleet_default, write=True
    )
    try:
        return await service.revert_to_version(
            db, layer, kind, name, data.version_number, current_user.id
        )
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)
        ) from exc


@router.delete(
    "/{name}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete a unit so the next layer down applies again",
)
async def delete_agent_text_unit(
    name: str,
    kind: str = Query(...),
    organization_id: UUID | None = Query(None),
    fleet_default: bool = Query(False),
    db: AsyncSession = Depends(get_async_db),
    current_user: User = Depends(current_active_user),
    service: AgentTextUnitService = Depends(get_service),
) -> None:
    """Removes the addressed layer's row only — the embedded default is not one."""
    layer = await resolve_layer(
        db, current_user, organization_id, fleet_default, write=True
    )
    deleted = await service.delete_override(db, layer, kind, name)
    if not deleted:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Agent text unit not found: {kind}/{name}",
        )

File: frontend/src/lib/api/agent-text-units.ts
/**
 * Agent text-unit API — wraps the `/api/v1/agent-text-units` endpoint surface.
 *
 * Backs `/admin/coord/agent-commands` and `/admin/coord/agent-skills`. Both
 * routes read the SAME corpus: a unit is `(kind, name)` plus a `files` map of
 * relative path → text. `kind=command` is the degenerate single-file case
 * (`<name>.md`); `kind=skill` carries `SKILL.md` plus siblings. `kind` is a
 * widenable discriminator, not a two-value enum — `.claude/agents/*.md` is the
 * next unit with the same delivery gap.
 *
 * **Three layers resolve, and only two of them are stored here.** The runner
 * resolves `account override → fleet default → embedded default`:
 *
 * * `organization_id IS NOT NULL` — one account's override. Addressed by
 *   omitting `fleet_default` (the backend falls back to the caller's personal
 *   organization when `organization_id` is not supplied).
 * * `organization_id IS NULL` — the **fleet default**, inherited by every
 *   account that has not overridden the unit. Addressed with
 *   `fleet_default=true`. Reads are open; **writes require a superuser**,
 *   because one write there changes the whole fleet.
 * * the runner's **embedded** copy — compiled into the binary via
 *   `include_str!` and never uploaded, so it has no row anywhere in this API.
 *   A `DELETE` therefore removes a stored layer and lets the next one down
 *   apply; it never deletes a default.
 *
 * Response shapes are typed locally against the backend's
 * `app/api/v1/endpoints/agent_text_units.py` +
 * `app/services/agent_text_unit_service.py` Pydantic models — the same
 * local-interface precedent `agent-registry.ts` and `fleet.ts` follow, which
 * keeps the pages decoupled from the generated api-client snapshot regen.
 */

import { httpClient } from "@/services/service-factory";

const AGENT_TEXT_UNITS_API = "/api/v1/agent-text-units";

// =============================================================================
// Wire types (mirrors of the backend Pydantic models)
// =============================================================================

/** Relative path → text. Paths are validated at the corpus boundary. */
export type UnitFiles = Record<string, string>;

/**
 * One stored unit, from ONE layer.
 *
 * `source` says which layer this row came from: `"user"` for an account
 * override, `"fleet"` for the `organization_id IS NULL` default. It used to be
 * documented as always `"user"` — that was true only while the backend stored
 * no fleet layer at all, and it is **no longer true**: the list route returns
 * `"fleet"` rows for every unshadowed fleet default.
 */
export interface AgentTextUnit {
  id: string;
  organization_id: string | null;
  created_by_user_id: string | null;
  kind: string;
  name: string;
  files: UnitFiles;
  /** The path inside `files` holding the unit's primary text — `SKILL.md` for
   *  a skill, `<name>.md` otherwise. Server-computed; never guess it for a row
   *  you already hold. */
  entrypoint: string;
  checksum: string | null;
  is_shared: boolean;
  /** False = carried by the corpus but never offered to the harness as an
   *  invocable unit (the `_gate-registration` / `_loop-control` copy-source
   *  specs). The backend CHECK refuses `true` for an underscore-prefixed
   *  name, so the two can never disagree. */
  is_invocable: boolean;
  current_version: number;
  /** Which resolution LAYER this row was served from — `"user"` (account
   *  override) or `"fleet"` (the fleet default layer).
   *
   *  ⚠️ Unrelated to `source_path` / `source_commit` below despite the shared
   *  prefix: this names the layer, those name the config repo the text was
   *  imported from. Adjacent names, different concepts — the same warning the
   *  canonical Rust type and the backend model both carry. */
  source: string;
  /** Import provenance: the repo-relative path the text came from, e.g.
   *  `.claude/skills/coord-revive/`. `null` for text authored in the console —
   *  which includes text that WAS imported and has since been edited here,
   *  because a save clears provenance rather than let it go stale. */
  source_path: string | null;
  /** Import provenance: the full 40-char commit of the source repo, or `null`
   *  when no commit honestly describes the bytes (authored here, or imported
   *  from a dirty tree). Never an abbreviation and never a sentinel. */
  source_commit: string | null;
  created_at: string;
  updated_at: string;
}

/** One immutable row of the append-only version chain. */
export interface AgentTextUnitVersion {
  id: string;
  agent_text_unit_id: string;
  version_number: number;
  files: UnitFiles;
  checksum: string | null;
  created_by_user_id: string | null;
  change_description: string | null;
  /** Set when this version was written by a revert: the version its files were
   *  copied from. `null` for an ordinary edit. */
  restored_from: number | null;
  created_at: string;
}

export interface Pagination {
  total: number;
  limit: number;
  offset: number;
  has_more: boolean;
}

export interface AgentTextUnitListResponse {
  items: AgentTextUnit[];
  pagination: Pagination;
}

export interface AgentTextUnitVersionListResponse {
  items: AgentTextUnitVersion[];
  pagination: Pagination;
}

/** Body of the upsert (`POST ""`). Keyed server-side on
 *  `(layer, kind, name)`; every call appends a version. */
export interface AgentTextUnitCreate {
  kind: string;
  name: string;
  files: UnitFiles;
  change_description?: string | null;
  is_shared?: boolean;
  is_invocable?: boolean;
  /** Import provenance, for an importing client only. The console never sends
   *  these, and that is exactly why a console save CLEARS whatever an earlier
   *  import recorded: once the text has been edited here it is no longer a copy
   *  of that path at that commit, and a stale provenance is worse than none. */
  source_path?: string | null;
  source_commit?: string | null;
}

/** Body of the patch (`PATCH "/{name}"`). A `files` change appends a version;
 *  the other fields do not. */
export interface AgentTextUnitUpdate {
  files?: UnitFiles;
  change_description?: string | null;
  is_shared?: boolean;
  is_invocable?: boolean;
}

// =============================================================================
// Layer addressing
// =============================================================================

/**
 * Which stored layer a request addresses.
 *
 * `fleetDefault: true` targets `organization_id IS NULL`. Otherwise the
 * request targets an account: `organizationId` names it explicitly, and
 * omitting it makes the backend fall back to the caller's personal
 * organization — the normal path for a single-user account, so we only send
 * the parameter when one is chosen.
 */
export interface LayerRef {
  organizationId?: string | null;
  fleetDefault?: boolean;
}

function layerParams(layer: LayerRef | undefined, params: URLSearchParams) {
  if (!layer) return;
  if (layer.fleetDefault) params.set("fleet_default", "true");
  if (layer.organizationId) params.set("organization_id", layer.organizationId);
}

function url(path: string, params: URLSearchParams): string {
  const query = params.toString();
  return query ? `${path}?${query}` : path;
}

// =============================================================================
// Requests
// =============================================================================

export interface ListUnitsOptions extends LayerRef {
  /** Filter to one kind, e.g. `"command"`. Omit to list every kind. */
  kind?: string;
  /**
   * Include fleet defaults the account has not overridden — i.e. ask the
   * server for the RESOLVED view rather than the raw layer.
   *
   * The console passes `false` and fetches the two layers separately, because
   * the resolved view cannot answer "does a fleet default exist *behind* this
   * override?" — the shadowed row is exactly what it drops.
   */
  includeFleetDefaults?: boolean;
  /**
   * Drop the non-invocable units — the underscore-prefixed copy-source specs
   * (`_gate-registration`, `_loop-control`).
   *
   * **The console must leave this false, and does.** A client that PROVISIONS
   * units to disk has to pass `true`, because a `_gate-registration.md`
   * written into `.claude/commands/` becomes an invocable
   * `/_gate-registration`. A client that EDITS the corpus is the opposite
   * case: the specs are real corpus members, and hiding them would leave an
   * operator unable to see or fix the text other units paste from. It is
   * declared here so the distinction is visible at the call site rather than
   * rediscovered.
   */
  invocableOnly?: boolean;
  offset?: number;
  limit?: number;
}

/** `GET /api/v1/agent-text-units` — one layer, or the resolved view. */
export async function listAgentTextUnits(
  options: ListUnitsOptions = {}
): Promise<AgentTextUnitListResponse> {
  const params = new URLSearchParams();
  if (options.kind) params.set("kind", options.kind);
  if (options.includeFleetDefaults !== undefined) {
    params.set("include_fleet_defaults", String(options.includeFleetDefaults));
  }
  if (options.invocableOnly !== undefined) {
    params.set("invocable_only", String(options.invocableOnly));
  }
  if (options.offset !== undefined)
    params.set("offset", String(options.offset));
  if (options.limit !== undefined) params.set("limit", String(options.limit));
  layerParams(options, params);
  return httpClient.get<AgentTextUnitListResponse>(
    url(AGENT_TEXT_UNITS_API, params)
  );
}

/** `POST /api/v1/agent-text-units` — create or replace a unit in one layer. */
export async function upsertAgentTextUnit(
  data: AgentTextUnitCreate,
  layer?: LayerRef
): Promise<AgentTextUnit> {
  const params = new URLSearchParams();
  layerParams(layer, params);
  return httpClient.post<AgentTextUnit>(
    url(AGENT_TEXT_UNITS_API, params),
    data
  );
}

/** `PATCH /api/v1/agent-text-units/{name}` — update an existing unit. */
export async function updateAgentTextUnit(
  kind: string,
  name: string,
  data: AgentTextUnitUpdate,
  layer?: LayerRef
): Promise<AgentTextUnit> {
  const params = new URLSearchParams({ kind });
  layerParams(layer, params);
  return httpClient.patch<AgentTextUnit>(
    url(`${AGENT_TEXT_UNITS_API}/${encodeURIComponent(name)}`, params),
    data
  );
}

/** `GET /api/v1/agent-text-units/{name}/versions` — newest version first.
 *
 *  History belongs to the ADDRESSED LAYER only: asking for an account's
 *  history never falls back to the fleet default's chain, because they are
 *  different rows with different edits. */
export async function listAgentTextUnitVersions(
  kind: string,
  name: string,
  layer?: LayerRef
): Promise<AgentTextUnitVersionListResponse> {
  const params = new URLSearchParams({ kind });
  layerParams(layer, params);
  return httpClient.get<AgentTextUnitVersionListResponse>(
    url(`${AGENT_TEXT_UNITS_API}/${encodeURIComponent(name)}/versions`, params)
  );
}

/**
 * `POST /api/v1/agent-text-units/{name}/revert` — APPENDS a new head whose
 * files equal the target version's. It never rewinds the chain, so the
 * response carries a NEW `current_version`, not `version_number`.
 */
export async function revertAgentTextUnit(
  kind: string,
  name: string,
  versionNumber: number,
  layer?: LayerRef
): Promise<AgentTextUnit> {
  const params = new URLSearchParams({ kind });
  layerParams(layer, params);
  return httpClient.post<AgentTextUnit>(
    url(`${AGENT_TEXT_UNITS_API}/${encodeURIComponent(name)}/revert`, params),
    { version_number: versionNumber }
  );
}

/**
 * `DELETE /api/v1/agent-text-units/{name}` — removes the addressed layer's row
 * so the next layer down applies again (204 No Content).
 *
 * ⚠️ The version chain goes WITH it: `AgentTextUnitVersion.agent_text_unit_id`
 * is `ondelete="CASCADE"` and the ORM relationship is
 * `cascade="all, delete-orphan"`, so this permanently discards every stored
 * version of that layer's row. Callers must say so before firing.
 */
export async function deleteAgentTextUnit(
  kind: string,
  name: string,
  layer?: LayerRef
): Promise<void> {
  const params = new URLSearchParams({ kind });
  layerParams(layer, params);
  await httpClient.delete<void>(
    url(`${AGENT_TEXT_UNITS_API}/${encodeURIComponent(name)}`, params)
  );
}
